lib/msf/core/db_manager/exploit_attempt.rb

```diff
@@ -27,186 +27,145 @@ def report_exploit_attempt(host, opts)
   end
 
+  # Create an `Mdm::ExploitAttempt` (and possibly an `Mdm::VulnAttempt`, if
+  # the `vuln` option is passed).
+  #
+  # @return (see #do_report_failure_or_success)
   def report_exploit_success(opts)
-  ::ActiveRecord::Base.connection_pool.with_connection {
-
-    wspace = opts.delete(:workspace) || workspace
-    mrefs  = opts.delete(:refs) || return
-    host   = opts.delete(:host)
-    port   = opts.delete(:port)
-    prot   = opts.delete(:proto)
-    svc    = opts.delete(:service)
-    vuln   = opts.delete(:vuln)
+    return unless opts[:refs]
+    host   = opts[:host] || return
 
-    timestamp = opts.delete(:timestamp)
-    username  = opts.delete(:username)
-    mname     = opts.delete(:module)
-
-    # Look up or generate the host as appropriate
-    if not (host and host.kind_of? ::Mdm::Host)
-      if svc.kind_of? ::Mdm::Service
-        host = svc.host
-      else
-        host = report_host(:workspace => wspace, :address => host )
-      end
-    end
-
-    # Bail if we dont have a host object
-    return if not host
+    wspace = opts[:workspace] || workspace
+    port   = opts[:port]
+    prot   = opts[:proto] || "tcp"
+    svc    = opts[:service]
 
     # Look up or generate the service as appropriate
     if port and svc.nil?
-      svc = report_service(:workspace => wspace, :host => host, :port => port, :proto => prot ) if port
+      opts[:proto] ||= "tcp"
+      opts[:service] = report_service(
+        workspace: wspace, host: host, port: port, proto: prot
+      )
     end
 
-    if not vuln
-      # Create a references map from the module list
-      ref_objs = ::Mdm::Ref.where(:name => mrefs.map { |ref|
-        if ref.respond_to?(:ctx_id) and ref.respond_to?(:ctx_val)
-          "#{ref.ctx_id}-#{ref.ctx_val}"
-        else
-          ref.to_s
-        end
-      })
+    do_report_failure_or_success(opts)
+  end
 
-      # Try find a matching vulnerability
-      vuln = find_vuln_by_refs(ref_objs, host, svc)
-    end
+  private
+
+  # @option opts [Array<String>, Array<Msf::Module::Reference>] :refs
+  # @option opts [Mdm::Host] :host
+  # @option opts [Mdm::Service] :service
+  # @option opts [Integer] :port (nil)
+  # @option opts ["tcp","udp"] :proto (nil)
+  # @option opts [Mdm::Vuln] :vuln (nil)
+  # @option opts [Time] :timestamp (nil)
+  # @option opts [Mdm::Vuln] :timestamp (nil)
+  # @option opts [String] :module (nil)
+  # @return [void]
+  def do_report_failure_or_success(opts)
+    ::ActiveRecord::Base.connection_pool.with_connection {
+      mrefs  = opts.delete(:refs) || return
+      host   = opts.delete(:host)
+      port   = opts.delete(:port)
+      prot   = opts.delete(:proto)
+      svc    = opts.delete(:service)
+      vuln   = opts.delete(:vuln)
+
+      timestamp  = opts.delete(:timestamp)
+      freason    = opts.delete(:fail_reason)
+      fdetail    = opts.delete(:fail_detail)
+      username   = opts.delete(:username)
+      mname      = opts.delete(:module)
+
+      if not vuln
```

Comment:
Since there's not `else`, this should use `unless vuln` or change to `if not vuln.nil?` if you want an exact test.